File: landrop-main/sender/desktop/src/protocol/transport.test.ts
import { describe, expect, it } from "vitest";

import { createInMemoryDuplex } from "./transport";

const text = (s: string): Uint8Array => new TextEncoder().encode(s);

describe("InMemoryTransport duplex", () => {
  it("delivers writes to the peer in order", () => {
    const { left, right } = createInMemoryDuplex();
    expect(left.connect()).toEqual({ ok: true });
    expect(right.connect()).toEqual({ ok: true });

    expect(left.write(text("aa"))).toEqual({ ok: true });
    expect(left.write(text("bb"))).toEqual({ ok: true });

    const first = right.read(16);
    expect(first).toMatchObject({ kind: "data" });
    if (first.kind === "data") {
      expect(new TextDecoder().decode(first.bytes)).toBe("aabb");
    }
    expect(right.read(16)).toEqual({ kind: "empty" });
  });

  it("returns empty while open with no data and end after the peer closes", () => {
    const { left, right } = createInMemoryDuplex();
    expect(right.read(4)).toEqual({ kind: "empty" });
    left.close();
    expect(left.write(text("x"))).toMatchObject({ ok: false });
    expect(right.read(4)).toEqual({ kind: "end" });
  });

  it("drains buffered data before reporting end", () => {
    const { left, right } = createInMemoryDuplex();
    left.write(text("payload"));
    left.close();
    expect(right.read(64)).toMatchObject({ kind: "data" });
    expect(right.read(64)).toEqual({ kind: "end" });
  });

  it("respects the maxBytes bound and preserves bytes across reads", () => {
    const { left, right } = createInMemoryDuplex();
    left.write(text("0123456789"));
    expect(right.read(4)).toMatchObject({ kind: "data" });
    const rest = right.read(64);
    expect(rest).toMatchObject({ kind: "data" });
    if (rest.kind === "data") {
      expect(new TextDecoder().decode(rest.bytes)).toBe("456789");
    }
  });

  it("copies writes so the caller may reuse its buffer", () => {
    const { left, right } = createInMemoryDuplex();
    const buf = text("abc");
    left.write(buf);
    buf[0] = 0x58; // 'X' — must not affect the transported copy
    const got = right.read(3);
    expect(got).toMatchObject({ kind: "data" });
    if (got.kind === "data") {
      expect(new TextDecoder().decode(got.bytes)).toBe("abc");
    }
  });

  it("io errors against the closed local end", () => {
    const { left, right } = createInMemoryDuplex();
    left.write(text("x"));
    left.close();
    expect(left.read(4)).toMatchObject({ kind: "error" });
    expect(left.read(4)).toMatchObject({ kind: "error" });
    expect(left.write(text("y"))).toMatchObject({ ok: false });
    expect(left.connect()).toMatchObject({ ok: false });
    expect(left.isClosed).toBe(true);
    // The peer still sees data, then end.
    expect(right.read(4)).toMatchObject({ kind: "data" });
    expect(right.read(4)).toEqual({ kind: "end" });
  });

  it("rejects a non-positive maxBytes", () => {
    const { left } = createInMemoryDuplex();
    expect(left.read(0)).toMatchObject({ kind: "error" });
    expect(left.read(-1)).toMatchObject({ kind: "error" });
  });

  it("close is idempotent", () => {
    const { left, right } = createInMemoryDuplex();
    left.close();
    left.close();
    expect(left.isClosed).toBe(true);
    right.close();
    right.close();
    // Reading on a locally-closed transport is an IO error, not EOF.
    expect(right.read(1)).toMatchObject({ kind: "error" });
  });
});

File: landrop-main/receiver/android/app/src/test/java/com/landrop/receiver/protocol/TransportTest.kt
package com.landrop.receiver.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class TransportTest {

    private fun text(s: String) = s.toByteArray(Charsets.UTF_8)

    private fun data(read: TransportRead): ByteArray =
        (read as? TransportRead.Data)?.bytes ?: error("expected Data, found $read")

    @Test
    fun deliversWritesToThePeerInOrder() {
        val (left, right) = createInMemoryDuplex()
        assertEquals(TransportResult.Ok, left.connect())
        assertEquals(TransportResult.Ok, right.connect())

        assertEquals(TransportResult.Ok, left.write(text("aa")))
        assertEquals(TransportResult.Ok, left.write(text("bb")))

        assertEquals("aabb", String(data(right.read(16)), Charsets.UTF_8))
        assertEquals(TransportRead.Empty, right.read(16))
    }

    @Test
    fun emptyWhileOpenAndEndAfterPeerCloses() {
        val (left, right) = createInMemoryDuplex()
        assertEquals(TransportRead.Empty, right.read(4))
        left.close()
        assertTrue(left.write(text("x")) is TransportResult.Failure)
        assertEquals(TransportRead.End, right.read(4))
    }

    @Test
    fun drainsBufferedDataBeforeReportingEnd() {
        val (left, right) = createInMemoryDuplex()
        left.write(text("payload"))
        left.close()
        assertEquals("payload", String(data(right.read(64)), Charsets.UTF_8))
        assertEquals(TransportRead.End, right.read(64))
    }

    @Test
    fun respectsMaxBytesAndPreservesBytesAcrossReads() {
        val (left, right) = createInMemoryDuplex()
        left.write(text("0123456789"))
        assertEquals("0123", String(data(right.read(4)), Charsets.UTF_8))
        assertEquals("456789", String(data(right.read(64)), Charsets.UTF_8))
    }

    @Test
    fun copiesWritesSoTheCallerMayReuseItsBuffer() {
        val (left, right) = createInMemoryDuplex()
        val buf = text("abc")
        left.write(buf)
        buf[0] = 'X'.code.toByte() // must not affect the transported copy
        assertEquals("abc", String(data(right.read(3)), Charsets.UTF_8))
    }

    @Test
    fun ioErrorsAgainstTheClosedLocalEnd() {
        val (left, right) = createInMemoryDuplex()
        left.write(text("x"))
        left.close()
        assertTrue(left.read(4) is TransportRead.Error)
        assertTrue(left.write(text("y")) is TransportResult.Failure)
        assertTrue(left.connect() is TransportResult.Failure)
        assertTrue(left.isClosed)
        // The peer still sees data, then end.
        assertEquals("x", String(data(right.read(4)), Charsets.UTF_8))
        assertEquals(TransportRead.End, right.read(4))
    }

    @Test
    fun rejectsANonPositiveMaxBytes() {
        val (left, _) = createInMemoryDuplex()
        assertTrue(left.read(0) is TransportRead.Error)
        assertTrue(left.read(-1) is TransportRead.Error)
    }

    @Test
    fun closeIsIdempotent() {
        val (left, right) = createInMemoryDuplex()
        left.close()
        left.close()
        assertTrue(left.isClosed)
        right.close()
        right.close()
        assertTrue(right.isClosed)
        // Reading on a locally-closed transport is an error, not EOF.
        assertTrue(right.read(1) is TransportRead.Error)
    }
}